test/export/test_export.py

```diff
@@ -7039,8 +7082,118 @@ def forward(self, x, y):
             failing_shapes=[
                 ((22,), (5,)),
             ],
+            test_serdes=True,
+        )
+
+    def test_dynamic_shapes_serdes_generic(self):
+        from torch.export.dynamic_shapes import Dim, _dumps, _extracts, _loads
+
+        class Foo(torch.nn.Module):
+            def forward(self, a, b, c, d):
+                if d == "hello":
+                    x = a[0] + a[1][1:]
+                    b = torch.cat([b, b], dim=0).reshape([-1, 1])
+                    return x + b, c * 2
+
+        # test de/serialization on some generic specs
+        dz = Dim("dz", min=4, max=16)
+        dx = 2*dz
+        dy = dx + 1
+        inputs = (
+            [
+                torch.randn(8, 4),
+                torch.randn(9, 4),
+            ],
+            torch.randn(4),
+            torch.randn(4, 4),
+            "hello",
+        )
+        dynamic_shapes = {
+            "a": [
+                (dx, 4),
+                (dy, 4),
+            ],
+            "b": (dz,),
+            "c": None,
+            "d": None,
+        }
+        ep = export(Foo(), inputs, dynamic_shapes=dynamic_shapes)
+        self._check_dynamic_shapes_specs_and_shapes(
+            Foo(),
+            inputs,
+            [dynamic_shapes],
+            [
+                ([(16, 4), (17, 4)], (8,), (4, 4), "hello"),
+                ([(24, 4), (25, 4)], (12,), (4, 4), "hello"),
+            ],
+            [
+                ([(16, 4), (17, 4)], (8,), (5, 5), "hello"),
+            ],
+            test_serdes=True,
+        )
+        self.assertExpectedInline(
+            _dumps(dynamic_shapes, inputs),
+            """{'dynamic_shapes': ([['2*dz', 4], ['2*dz + 1', 4]], ['dz'], [None, None], None), 'dims': {'dz': {'min': 4, 'max': 16, 'derived': ['2*dz', '2*dz + 1']}}}"""
+        )
+        self.assertExpectedInline(
+            _extracts(ep),
+            """{'dynamic_shapes': ([['2*d0', 4], ['2*d0 + 1', 4]], ['d0'], [4, 4], None), 'dims': {'d0': {'min': 4, 'max': 16, 'derived': ['2*d0', '2*d0 + 1']}}}""",
```

Comment:
`_extracts()` just enumerates the symbols, doing this as a placeholder if anyone has better ideas. Could use the placeholder names + indices.